### ISSUE TITLE
 [FEATURE]: Add support for logs and subroute modules

### ISSUE THREAD

**Issue by PaTTeeL** [mholt/caddy-l4#50]
Should this plugin support logs and subroute modules ?
My configuration like this:
```
  "apps":{
    "layer4":{"servers":{"stream":{
      "listen":[":443"],
      "logs":{"default_logger_name":"log0"},
      "routes":[
        { "terminal":true,
          "match":[{"tls":{"sni":["MyDomain.com"]}}],
          "handle":[{"handler":"subroute","routes":[
          { "handle":[{"handler":"tls"}]},
          { "handle":[{"handler":"proxy","upstreams":[{"dial":["unix//dev/shm/websocket"]}],"proxy_protocol":"v2"}],
            "match":[{"http":[{"header":{"Upgrade":["websocket"]},"path":["/websocket"]}]}]},
          { "handle":[{"handler":"proxy","upstreams":[{"dial":["local:80"]}]}]}]}]}]
    }}},
    "http":{"servers":{
      "srv1":{
        "allow_h2c":true,
        "listen":[":80"],
        "logs":{"default_logger_name":"log1"},
```

error log:

```
Mar 27 04:55:27  caddy[15504]: 2022/03/26 20:55:27.015        ERROR        watcher        applying latest config        {"config_file": "/etc/caddy/Caddyfile.json", "error": "loading new config: loading layer4 app module: decoding module config: layer4: json: unknown field \"logs\""}

Mar 27 04:57:50  caddy[15504]: 2022/03/26 20:57:50.015        ERROR        watcher        applying latest config        {"config_file": "/etc/caddy/Caddyfile.json", "error": "loading new config: loading layer4 app module: decoding module config: layer4: json: unknown field \"terminal\""}

Mar 27 04:58:33  caddy[15504]: 2022/03/26 20:58:33.015        ERROR        watcher        applying latest config        {"config_file": "/etc/caddy/Caddyfile.json", "error": "loading new config: loading layer4 app module: provision layer4: server 'srv0': route 0: position 0: loading module 'subroute': unknown module: layer4.handlers.subroute"}
```

Logs module makes the layer4 plugin output pretty logs by apart logs from program's kernel log .

And will subroute make caddy more secure by filter sni before actions ?

**Comment by RussellLuo** [comment]
There has already been an issue for supporting `subroute`, see #20. Maybe I can try to implement this if there's no other volunteer :)

**Comment by mholt** [comment]
> And will subroute make caddy more secure by filter sni before actions ?

No, subrouting doesn't add any security. It is just convenient for reducing repetition in some cases by nesting certain routing logic. You can do SNI routing without subroutes. They are orthogonal features.

**Comment by PaTTeeL** [comment]
> > And will subroute make caddy more secure by filter sni before actions ?
> 
> No, subrouting doesn't add any security. It is just convenient for reducing repetition in some cases by nesting certain routing logic. You can do SNI routing without subroutes. They are orthogonal features.
> 
I mean routes module can only process handles parallelly. So with subroute module, layer4 plugin can filter wrong or empty sni without leak certs, and stream current one to different servers by uri or headers which I couldn't find the way routes can be done.


> As for logs, what do you want logged that isn't already logged?

By now, layer4's log is mixed with admin's log, which like this `journalctl -u caddy`
```
Mar 27 08:52:17  caddy[15504]: 2022/03/27 00:52:17.011        INFO        watcher        config file changed; reloading        {"config_file": "/etc/caddy/Caddyfile.json"}
Mar 27 08:52:17  caddy[15504]: 2022/03/27 00:52:17.012        INFO        using provided configuration        {"config_file": "/etc/caddy/Caddyfile.json", "config_adapter": ""}
Mar 27 08:52:17  caddy[15504]: 2022/03/27 00:52:17.017        WARN        admin        admin endpoint disabled
Mar 27 08:52:17  caddy[15504]: 2022/03/27 00:52:17.019        WARN        http        server is listening only on the HTTP port, so no automatic HTTPS will be applied to this server        {"server_name": "srv1", "http_port": 80}
Mar 27 08:52:17  caddy[15504]: 2022/03/27 00:52:17.020        DEBUG        tls.cache        added certificate to cache        {"subjects": ["Domain.com"], "expiration": "2022/06/02 23:59:59.000", "managed": true, "issuer_key": "acme.zerossl.com-v2-DV90", "hash": "xxxxxxxxxxxxxxx", "cache_size": 1, "cache_capacity": 10000}
Mar 27 08:52:17  caddy[15504]: 2022/03/27 00:52:17.020        DEBUG        layer4        listening        {"address": "tcp/[::]:443"}
Mar 27 08:52:17  caddy[15504]: 2022/03/27 00:52:17.020        DEBUG        http        starting server loop        {"address": "[::]:80", "http3": false, "tls": false}
Mar 27 08:52:17  caddy[15504]: 2022/03/27 00:52:17.021        INFO        tls.cache.maintenance        started background certificate maintenance        {"cache": "0xc000111ab0"}
Mar 27 08:52:17  caddy[15504]: 2022/03/27 00:52:17.024        INFO        tls.cache.maintenance        stopped background certificate maintenance        {"cache": "0xc000110700"}
Mar 27 08:52:17  caddy[15504]: 2022/03/27 00:52:17.024        INFO        autosaved config (load with --resume flag)        {"file": "/var/lib/caddy/.config/caddy/autosave.json"}
Mar 27 08:52:37  caddy[15504]: 2022/03/27 00:52:37.070        DEBUG        tls.handshake        choosing certificate        {"identifier": "Domain.com", "num_choices": 1}
Mar 27 08:52:37  caddy[15504]: 2022/03/27 00:52:37.071        DEBUG        tls.handshake        default certificate selection results        {"identifier": "Domain.com", "subjects": ["Domain.com"], "managed": true, "issuer_key": "acme.zerossl.com-v2-DV90", "hash": "xxxxxxxxxxxxxxxxxxxxx"}
Mar 27 08:52:37  caddy[15504]: 2022/03/27 00:52:37.071        DEBUG        tls.handshake        matched certificate in cache        {"subjects": ["Domain.com"], "managed": true, "expiration": "2022/06/02 23:59:59.000", "hash": "xxxxxxxxxxxxxxxxxxxxxxxxxxxx"}
Mar 27 08:52:37  caddy[15504]: 2022/03/27 00:52:37.515        DEBUG        layer4.handlers.tls        terminated TLS        {"remote": "remote_ip:port", "server_name": "Domain.com"}
Mar 27 08:52:37  caddy[15504]: 2022/03/27 00:52:37.516        DEBUG        layer4.handlers.proxy        dial upstream        {"remote": "remote_ip:port", "upstream": "127.0.0.1:80"}
Mar 27 08:52:37  caddy[15504]: 2022/03/27 00:52:37.918        DEBUG        http.handlers.reverse_proxy        selected upstream        {"dial": "127.0.0.1:8005", "total_upstreams": 1}
Mar 27 08:52:37  caddy[15504]: 2022/03/27 00:52:37.927        DEBUG        http.handlers.reverse_proxy        upstream roundtrip        {"upstream": "127.0.0.1:8005", "duration": 0.008124022, "request": {"remote_ip": "127.0.0.0", "remote_port": "47644", "proto": "HTTP/2.0", "method": "POST", "host": "MyDomain.com", "uri": "f85c72d0", "headers": {"X-Forwarded-Proto": ["http"], "X-Forwarded-Host": "MyDomain.com", "Content-Type": ["application/grpc"], "User-Agent": ["grpc-go/1.44.1-dev"], "Te": ["trailers"], "X-Forwarded-For": "remote_ip"}}, "headers": {"Content-Type": ["application/grpc"]}, "status": 200}
```

So if layer4's log can output like configurations below:
```
  "logging":{"logs":{
    "default":{
      "level":"DEBUG","encoder":{"format":"console","level_format":"upper","time_format":"wall_milli"},
      "exclude":["http.log.access.log0","layer4.log.access.stream0"]},
    "log0":{
      "level":"DEBUG","encoder":{"format":"json","level_format":"upper","time_format":"wall_milli"},
      "writer":{"filename":"/tmp/caddy.log","output":"file"},
      "include":["http.log.access.log0","layer4.log.access.stream0"]}}},
  "apps":{
    "http":{"servers":{
      "srv0":{
        "listen":[":80"],
        "logs":{"default_logger_name":"log0"},
    "layer4":{"servers":{"stream0":{
      "listen":[":443"],
      "logs":{"default_logger_name":"stream0"},
```
Makes all access actions together from layer4 to https in one file "caddy.log". And keeps caddy's admin log in system journal.
But it seems http.handlers.*(like http.handlers.file_server , http.handlers.reverse_proxy , http.handlers.rewrite and http.log.error.log1 ) still in system journal even I've defined `"writer":{"filename":"/tmp/caddy.log","output":"file"},"include":["http.log.access.log0"]` ?

**Comment by francislavoie** [comment]
`include` and `exclude` map directly to the logger name, so if you want to include `http.handlers.*` in another logger, then put `http.handler` in that logger.

`layer4` doesn't have access logs right now, but you can just move layer4 logs to another logger by including `layer4`, etc.

**Comment by PaTTeeL** [comment]
> `include` and `exclude` map directly to the logger name, so if you want to include `http.handlers.*` in another logger, then put `http.handler` in that logger's `include`.
> 
> `layer4` doesn't have access logs right now, but you can just move layer4 logs to another logger by including `layer4`, etc.

It works !
Thank you so much for help!
Now I splitted http and layer4's log from tls and admin's log by `"include":["http","layer4"]` and `"exclude":["http","layer4"]`

**Comment by mholt** [comment]
> I mean routes module can only process handles parallelly.

Can you explain what you mean by this? Each handler in a route is called sequentially in turn, not in parallel. And each route is evaluated sequentially as well.

(Thanks Francis, for answering questions about logging.)

**Comment by PaTTeeL** [comment]
> 
> Can you explain what you mean by this? Each handler in a route is called sequentially in turn, not in parallel. And each route is evaluated sequentially as well.
> 
> (Thanks Francis, for answering questions about logging.)

> 

Sorry for my pool expression.
I mean with subroute, I can upstream with both sni & uri, but now, routes module can only choose one matcher which either tls's sni or http's header.
Seems a graph maybe clear: 
Route:
```
|
----->If SNI: domain A------>Then Dial: Server1
|
----->If SNI: domain B------->Then Dial: Server2
|
----->If HTTP Header: URI A------>Then Dial: Server3
|
----->If HTTP Header: URI B------>Then Dial: Server4
```
parallelly shoud be they have the same priority?

Subrouter:
```
|
------>If SNI: domain A
                       |
                       -----------IF HTTP Header: URI A------>Then Dial: Server1
                       |
                       -----------IF HTTP Header: URI B------>Then Dial: Server2
|
------>If SNI: domain B
                       |
                       -----------IF HTTP Header: URI A------>Then Dial: Server3
                       |
                       -----------IF HTTP Header: URI B------>Then Dial: Server4
```

**Comment by francislavoie** [comment]
You can have more than one matcher for each route, so you can do "if SNI && if HTTP" in one matcher.

Either way, I'm pretty sure we can close this as a duplicate of https://github.com/mholt/caddy-l4/issues/20 at this point, we know that you're looking for `layer4.handlers.subroute` which that issue asks for, and your log question has been answered.

**Comment by PaTTeeL** [comment]
> You can have more than one matcher for each route, so you can do "if SNI && if HTTP" in one matcher.
> 
May I ask how to write configurations for && in route ? I've searched for days.

> Either way, I'm pretty sure we can close this as a duplicate of #20 at this point, we know that you're looking for `layer4.handlers.subroute` which that issue asks for, and your log question has been answered.

**Comment by francislavoie** [comment]
This is an AND:
```json
"match": [
	{
		"sni": [],
		"http": []
	}
]
```

This is an OR:
```json
"match": [
	{
		"sni": []
	},
	{
		"http": []
	}
]
```

**Comment by PaTTeeL** [comment]
> This is an AND:
> 
> ```json
> "match": [
> 	{
> 		"tls": {},
> 		"http": []
> 	}
> ]
> ```
> 
> This is an OR:
> 
> ```json
> "match": [
> 	{
> 		"tls": {}
> 	},
> 	{
> 		"http": []
> 	}
> ]
> ```
> 
> The layer4 `match` docs on caddyserver.com don't seem to mention this, but the HTTP app does it the same way, and explains:
> 
> https://caddyserver.com/docs/json/apps/http/servers/routes/match/
> 
> > The matcher sets which will be used to qualify this route for a request (essentially the "if" statement of this route). Each matcher set is OR'ed, but matchers within a set are AND'ed together.
> 
> You can also read through the code to understand how it's implemented (it's decently commented) https://github.com/mholt/caddy-l4/blob/master/layer4/matchers.go
> 
> But I guess in this case you might _need_ a subroute because you'd need to use the `tls` handler to terminate TLS before you can match `http` things, I think.

Yes, I've read both
https://caddyserver.com/docs/json/apps/http/servers/routes/match/
https://caddyserver.com/docs/json/apps/http/servers/routes/handle/
and tried configurations like below while only server2 works.
```
        { "handle":[
            {"handler":"tls"},
            {"handler":"proxy","upstreams":[{"dial":["server1"]}],"proxy_protocol":"v2"}],
          "match":[{
            "tls":{"sni":["DomainA"]},
            "http":[{"path":["URI1"]}]}]},
        { "handle":[
            {"handler":"tls"},
            {"handler":"proxy","upstreams":[{"dial":["server2"]}],"proxy_protocol":"v2"}],
          "match":[
            {"tls":{"sni":["DomainB"]}},
            {"http":[{"path":["URI2"]}]}
          ]
        },
```
Matcher with DomainA&URI1 does not work, and all DomainB's connections was streamed to server2.


And there's another question: the layer4 module cannot reconizge grpc/http2 connections with similar configuration like below, but there's no upstream record in logs :
```
        { "handle":[{"handler":"tls"}]},
        { "handle":[
            {"handler":"proxy","upstreams":[{"dial":["127.0.0.1:8003"]}]}
          ],
          "match":[{"http":[{"header":{"Content-Type":["application/grpc*"]}}]}
          ]
        { "handle":[
            {"handler":"proxy","upstreams":[{"dial":["127.0.0.1:8003"]}]}
          ],
          "match":[{"http":[{"path":["/grpc"]}]}
          ]
        },
```
Both configurations does not work.
The only way to upstream gRPC is using layer4 to proxy all other connections (incl gRPC) to caddy server and then use reverse_proxy by match Content-Type or path.
example:
```
          { "handle":[{
              "handler":"reverse_proxy",
              "flush_interval":-1,
              "upstreams":[{"dial":gRPC Server"}],
              "transport":{"protocol":"http","versions":["h2c"]}}],
            "match":[{"header":{"Content-Type":["application/grpc*"]},"path":["/grpc"]}]},
```
So what's the difference between caddy server and layer4 in gRPC ?
May I upstream gRPC connections by using layer4 module?

**Comment by mholt** [comment]
I believe the same question is posted in a new issue #51, and with #52 merged, I'll consider this issue finished.
